src/LinearOperators.jl

```diff
@@ -85,346 +90,355 @@ end
 symmetric."""
-LinearOperator(M :: SymTridiagonal) =
-    LinearOperator(M; symmetric = true, hermitian = eltype(M) <: Real)
+LinearOperator{T}(M :: SymTridiagonal{T}) =
+    LinearOperator(M; symmetric=true, hermitian=eltype(M) <: Real)
 
+# the only advantage of this constructor is optional args
+# use LinearOperator{Float64} if you mean real instead of complex
 "Construct a linear operator from functions."
-LinearOperator(nrow :: Int, ncol :: Int, dtype :: DataType,
-               symmetric :: Bool, hermitian :: Bool,
-               prod :: Function, tprod :: Function, ctprod :: Function) =
-  LinearOperator(nrow, ncol, dtype, symmetric, hermitian,
-                 prod, Nullable{Function}(tprod), Nullable{Function}(ctprod))
-
-"Construct a real symmetric linear operator from a function."
-LinearOperator(nrow :: Int, dtype :: DataType, prod :: Function) =
-  LinearOperator(nrow, nrow, dtype, true, true,
-                 prod,
-                 Nullable{Function}(prod),
-                 Nullable{Function}(prod))
-
-"Construct a linear operator from a single function."
-LinearOperator(nrow :: Int, ncol :: Int, dtype :: DataType,
-               symmetric :: Bool, hermitian :: Bool,
-               prod :: Function) =
-  LinearOperator(nrow, ncol, dtype, symmetric, hermitian,
-                 prod, Nullable{Function}(), Nullable{Function}())
-
-
-if VERSION ≥ v"0.4.0-dev"
-  import Base.+, Base.-, Base.*, Base.(.+), Base.(.-), Base.(.*)
-  import Base.transpose, Base.ctranspose
+function LinearOperator(nrow :: Int, ncol :: Int,
+                        symmetric :: Bool, hermitian :: Bool,
+                        prod :: Function,
+                        tprod :: Union{Function,Nullable{Function}}=Nullable{Function}(),
+                        ctprod :: Union{Function,Nullable{Function}}=Nullable{Function}())
+
+  T = hermitian ? (symmetric ? Float64 : Complex128) : Complex128
+  LinearOperator{T}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
 end
 
+# "Construct a real symmetric or complex hermitian linear operator from a function."
+# LinearOperator(nrow :: Int, T :: DataType, prod :: Function) =
+#   LinearOperator{T}(nrow, nrow, T <: Real, T <: Real || T <: Complex,
+#                     prod, Nullable{Function}(prod), Nullable{Function}(prod))
+
+# "Construct a linear operator from a single function."
+# LinearOperator(nrow :: Int, ncol :: Int,
+#                symmetric :: Bool, hermitian :: Bool,
+#                prod :: Function) =
+#   LinearOperator(nrow, ncol, symmetric, hermitian,
+#                  prod, Nullable{Function}(), Nullable{Function}())
+
+
 # Apply an operator to a vector.
-function (*)(op :: AbstractLinearOperator, v :: Vector)
-  (m, n) = size(op)
-  if size(v,1) != n
-    error("Shape mismatch")
-  end
-  return op.prod(v)
+function *(op :: AbstractLinearOperator, v :: Vector)
+  size(v, 1) == size(op, 2) || throw(LinearOperatorException("shape mismatch"))
+  op.prod(v)
 end
```

Comment:
An operator with element type Float64 could very well be applied to a complex vector. What behavior do we want here? Should that be allowed? Should there be a warning?

The commented code below shows variants where we restrict `v` to having the same element type as the operator, and a more general multiplication where all mixtures are allowed. If the element type of `v` is "larger" than the operator's, the current solution is to create a new operator, which I don't like at all.

It seems to me that we evidently want to be able to multiply an operator with a vector of a "smaller" type (e.g., a Float64 operator with an Int vector; the result is a Float64 vector). But what about a Float64 operator with a Complex128 vector?